README.md
# w3x-my-tower
地图 - 我塔非凡

模式
```
1、百波模式通过100关，打败最后boss即可胜利
2、无尽模式，当中心的大佬被扣血为0时才失败
3、内战模式，请击败其他玩家，即可胜利并获得奖励（AI也是）
```

- 打怪捡石头，可以设置兵塔，也可以用来升级
- 得到的技能书，可以自由的使用、升级、替换
- 开局尽快设置4座辅助塔，可以帮助主塔攻防（游戏刚开始时兵塔石很容易获得）
- 5个兵塔可以组成种族联动，实现更强的队伍能力
- 天赋9级最高，越高基础能力越强，但升级能力不会变
- 对战模式有AI，开心去玩！一样有奖励

- 兵塔石除了可以设置塔还可以吃掉升级
- 天赋抽到9级之后，同一个单位不会洗掉
- 有些技能只有对战有大用，注意看物品说明和提示
- 种族的“队伍”效果根据情况你的能力会延时生效
- 辅助塔的天赋比主塔更给力
- 木头是免费的，每一局结束累计，使用后消失

此作拥有丰富的近200种以上兵种，1000个以上自定义技能，可以自由组合兵塔形成属于他的单位
玩家还可以自由使用兵塔石，学习、替换不同的技能书，达到一个属于自己的独特单位
更有天赋、物品升华、队伍羁绊效果，使得游戏有趣而多变，无论是初期还是后期都不会无聊
游戏欢乐的内战方式更是设有聪明的AI，对单人玩家非常友好，上手容易，自由探索，魅力十足！

木头
```
玩家可以获得结束游戏时的与波数相同的木头，对战模式胜利者奖励额外木头
```

称号
```
初出茅庐 总场>=0 胜场>=0
略有小成 总场>=3 胜场>=0
初露锋芒 总场>=5 胜场>=0
游刃有余 总场>=10 胜场>=0
一战成名 总场>=10 胜场>=1
名扬四方 总场>=10 胜场>=2
当世雄豪 总场>=20 胜场>=3
登峰造极 总场>=20 胜场>=5
猎尽天下 总场>=30 胜场>=7
超然世外 总场>=30 胜场>=9
超凡入圣 总场>=30 胜场>=11
身越七魄 总场>=40 胜场>=13
灵通三魂 总场>=50 胜场>=15
灭劫星窍 总场>=50 胜场>=25
神游三界 总场>=60 胜场>=50
六道大仙 总场>=80 胜场>=75
九天至尊 总场>=100 胜场>=100
```

每个玩家有一个兵塔和一个信使，兵塔就是攻击的单位，打怪和商店可以获得随机的兵塔类型，使用后可以更换兵塔。
兵塔类型由弱到强分为：E、D、C、B、A、S、SS、SSS级别，越高级成长越快,杀怪也会升级，越强力
有的兵塔类型也会特别强,即使是同为A级有的可能接近S的能力，有的仅仅比B强一点点
兵塔可以设置6个技能
信使就是运输单位，没有其他的作用，可以使用所有物品，例如技能书

DzData

lumber 木头
prestige 称号
prestigecs 局数
prestigess 胜场

激活方式：进入游戏自动生效创建召唤符在您的信使上，使用物品召唤信使
激活方式：进入游戏自动生效创建套装包在您的信使上，使用物品装扮套装

道具key必须大写
ICEMON 信使冰戟剑灵 信使冰戟剑灵，可以闪烁、可以发起技能附加冰攻击并令敌人移动减速
PHOENIX 信使涅槃火凤凰 信使涅槃火凤凰，可以闪烁、可以发起技能附加火攻击并增强火伤害
TZFIRE 炎炎焚烧套装 周身冒火热烈的特效套装，包含：炎炎旋风、烧灼圈圈、焚烧之火、鬼马火球
TZGOLD 金碧辉煌套装 带金翅膀和小精灵的光辉特效套装，包含：金耀翅膀、公正光辉、天堂圣音、金色精灵
TZBLOOD 血色炽热套装 赤红代表勇敢的特效套装，包含：血色翅膀、赤红漩涡、祭奠符文
TZDRAGON 青龙碧翼套装 青龙环绕的特效套装，包含：青空之翼、游龙欢悦、龙腾吐息
TZDARK 迷幻黑紫套装 宁静低调的暗黑翅膀特效套装，包含：幻黑翅膀、迷紫幻象、迷惑雾阵
TZGHOST 邪鬼怨灵套装 周身带有邪眼鬼魂，包含：邪鬼怨灵、邪鬼阵法、超度符文、逗趣鬼眼
TZSWORD 出云剑仙套装 环绕飞剑并带有白龙飞升的特效套装，包含：盘旋飞剑、灵剑气场、飞龙在天

CHANGE_LOG.txt
有建议和bug可以来[企鹅群325338043]反馈

# 1.0.4
平衡性调整

# 1.0.3
优化内存占用
修复异步问题
修复已知的bug
增加地图彩蛋

# 1.0.2
优化内存占用
优化异步问题
修正boss行为
修复已知的bug
现在随着波数增加boss会逐渐作出反制
略微调整对战AI
 - 更快速的决策
 - 更理智的升华
 - 更合理地敕令

# 1.0.1
优化内存占用
优化物品的描述
修复一些无效的技能，优化一些不和谐的图标
平衡性调整，百波简单化，无尽困难化
英雄技能增强，种族效果暂时微调
现在所有装备都固定可以叠加3次。
视距调整加入，输入+number或-number调整视距。
信使之笼的信使删除了一部分，然后现在可以开局自由替换了。
现在信使移动速递都改为522，并且拥有闪烁技能和区域拾取技能。
现在如果不是对战，不会掉落防御性技能，升华也是一样
现在天赋升级需要技能书，然后免费升级
增强了天赋对于攻击力的加成效果
略微调整对战AI：
- 现在AI现在出的物品更具针对性，并且学会叠加物品和升华物品
- 现在AI也会使用敕令战略

# 1.0.0
万恶的初版

init/towers.lua
-- towers
--兵塔变敌军单位
local createTowerShadowUnit = function(v,sti,tlv)
    local sobj = slk.unit.ogru:new("this_tower_shadow_" .. v.INDEX)
    sobj.Name = "["..tlv.."阶]" .. v.Name
    sobj.upgrades = ""
    sobj.file = v.file
    sobj.Art = v.Art
    sobj.modelScale = 1.00
    sobj.scale = 1.20
    sobj.HP = 100
    sobj.spd = 100
    sobj.sight = 1200
    sobj.nsight = 1200
    sobj.unitSound = v.unitSound or ""
    sobj.weapsOn = 1
    sobj.dmgplus1 = 1 -- 基础攻击
    sobj.showUI1 = 0 -- 不显示攻击按钮
    sobj.fused = 0
    sobj.canFlee = 0
    local targs1 = "vulnerable,ground,ward,structure,organic,mechanical,tree,debris,air" --攻击目标
    sobj.targs1 = targs1
    if (v.weapTp1 == "msplash" or v.weapTp1 == "artillery") then
        --溅射/炮火
        sobj.Farea1 = v.Farea1 or 1
        sobj.Qfact1 = v.Qfact1 or 0.05
        sobj.Qarea1 = v.Qarea1 or 500
        sobj.Hfact1 = v.Hfact1 or 0.15
        sobj.Harea1 = v.Harea1 or 350
        sobj.splashTargs1 = targs1 .. ",enemies"
    end
    if (v.weapTp1 == "mbounce") then
        --弹射
        sobj.Farea1 = v.Farea1 or 450
        sobj.targCount1 = v.targCount1 or 4
        sobj.damageLoss1 = v.damageLoss1 or 0.3
        sobj.splashTargs1 = targs1 .. ",enemies"
    end
    if (v.weapTp1 == "mline") then
        --穿透
        sobj.spillRadius = v.spillRadius or 200
        sobj.spillDist1 = v.spillDist1 or 450
        sobj.damageLoss1 = v.damageLoss1 or 0.3
        sobj.splashTargs1 = targs1 .. ",enemies"
    end
    if (v.weapTp1 == "aline") then
        --炮火穿透
        sobj.Farea1 = v.Farea1 or 1
        sobj.Qfact1 = v.Qfact1 or 0.05
        sobj.Qarea1 = v.Qarea1 or 500
        sobj.Hfact1 = v.Hfact1 or 0.15
        sobj.Harea1 = v.Harea1 or 350
        sobj.spillRadius = v.spillRadius or 200
        sobj.spillDist1 = v.spillDist1 or 450
        sobj.damageLoss1 = v.damageLoss1 or 0.3
        sobj.splashTargs1 = targs1 .. ",enemies"
    end
    sobj.acquire = 749 + (v.RangeInc or 0)
    sobj.backSw1 = v.backSw1 or 0.500
    sobj.dmgpt1 = v.dmgpt1 or 0.500
    sobj.rangeN1 = 750 + (v.RangeInc or 0)
    sobj.cool1 = v.cool1 or 2.00
    sobj.armor = "Flesh" -- 被击声音
    sobj.targType = "ground" --作为目标类型
    sobj.Missileart = v.Missileart -- 箭矢模型
    sobj.Missilespeed = 1100 -- 箭矢速度
    sobj.Missilearc = v.Missilearc or 0.05
    sobj.weapTp1 = v.weapTp1 or "normal" --攻击类型
    sobj.weapType1 = "" --攻击声音
    sobj.upgrades = ""
    sobj.Builds = ""
    local movetp = v.movetp or "foot"
    local moveHeight = v.moveHeight or 0
    if(movetp == 'fly')then
        moveHeight = 250
    end
    sobj.movetp = movetp --移动类型
    sobj.moveHeight = moveHeight --移动高度
    sobj.moveFloor = moveHeight * 0.25 --最低高度
    sobj.regenHP = 0
    sobj.regenType = ""
    sobj.def = 0
    local abl = {}
    if(type(v.abilList) == "string")then
        abl = string.explode(',', abl)
    elseif(type(v.abilList) == "table")then
        for _,t in pairs(v.abilList) do
            table.insert( abl, t )
        end
    end
    if(#abl ==0)then
        abl = {
            towerSpxKV["封印枷锁之一"],
            towerSpxKV["封印枷锁之二"]
        }
    elseif(#abl == 1)then
        table.insert( abl, towerSpxKV["封印枷锁之二"] )
    end
    table.insert( abl, v.DESC_ABILITY_ID )
    table.insert( abl, unitPowerMap[tlv] )
    sobj.abilList = string.implode(",",abl) .. "," .. string.implode(",",LINK_ABILITY_STACK)
    v.TOWER_ID = v.UNIT_ID
    v.TYPE = "tower_shadow"
    v.UNIT_ID = sobj:get_id()
    ?>
    call SaveStr(hash_myslk, StringHash("towers_shadow"), <?=sti?>, "<?=string.addslashes(json.stringify(v))?>")
    <?
end


towers = {
    E = towers_e,
    D = towers_d,
    C = towers_c,
    B = towers_b,
    A = towers_a,
    S = towers_s,
    SS = towers_ss,
    SSS = towers_sss
}
local towersTi = 0
for tlv, tow in pairs(towers) do
    if tow ~= nil then
        for i, v in pairs(tow) do
            towersTi = towersTi + 1
            local thisIndex = v.Name
            -- 处理塔基数据
            local UberDesc = v.Ubertip
            local Ubertip = ""
            local obj = slk.unit.Hpal:new("towers_" .. thisIndex)

            Primary = v.Primary or "STR"
            v.STR = v.STR or 1
            v.AGI = v.AGI or 1
            v.INT = v.INT or 1
            local tempPower = 0
            local tempAttack = 0
            if (tlv == "E") then
                tempPower = 20 + math.random(-2,3)
                tempAttack = 30 + math.random(-1,4)
            elseif (tlv == "D") then
                tempPower = 30 + math.random(-3,4)
                tempAttack = 60 + math.random(-2,6)
            elseif (tlv == "C") then
                tempPower = 45 + math.random(-4,6)
                tempAttack = 90 + math.random(-4,8)
            elseif (tlv == "B") then
                tempPower = 60 + math.random(-4,6)
                tempAttack = 150 + math.random(-6,10)
            elseif (tlv == "A") then
                tempPower = 80 + math.random(-5,8)
                tempAttack = 220 + math.random(-8,13)
            elseif (tlv == "S") then
                tempPower = 100 + math.random(-6,9)
                tempAttack = 350 + math.random(-10,15)
            elseif (tlv == "SS") then
                tempPower = 140 + math.random(-7,11)
                tempAttack = 450 + math.random(-15,20)
            elseif (tlv == "SSS") then
                tempPower = 180 + math.random(-8,15)
                tempAttack = 600 + math.random(-25,75)
            end
            if(Primary == "STR")then
                v.STR = math.floor(tempPower * 0.06 * math.random(7,11))
                v.AGI = math.floor(tempPower * 0.025 * math.random(5,8))
                v.INT = math.floor(tempPower * 0.025 * math.random(5,9))
                tempAttack = tempAttack + 19 * math.random()
            elseif(Primary == "AGI")then
                v.STR = math.floor(tempPower * 0.035 * math.random(5,8))
                v.AGI = math.floor(tempPower * 0.06 * math.random(6,10))
                v.INT = math.floor(tempPower * 0.025 * math.random(3,8))
                tempAttack = tempAttack + 25 * math.random()
            elseif(Primary == "INT")then
                v.STR = math.floor(tempPower * 0.03 * math.random(7,9))
                v.AGI = math.floor(tempPower * 0.02 * math.random(5,8))
                v.INT = math.floor(tempPower * 0.06 * math.random(8,11))
                tempAttack = tempAttack + 21 * math.random()
            end
            tempAttack = tempAttack * 1.4
            if(tempAttack < 0)then
                tempAttack = 0
            end
            v.STRplus = v.STR * 0.30
            v.AGIplus = v.AGI * 0.30
            v.INTplus = v.INT * 0.30
            v.ATTACK_WHITE = math.floor(tempAttack)
            v.RACE = string.explode("·",v.Name)[1]
            local ThreeTotal = v.STR + v.AGI + v.INT
            local coolMark = math.floor(100 / (v.cool1 or 2.00))
            local TowerMark = math.floor(
                v.STR + v.AGI + v.INT 
                + v.ATTACK_WHITE
                + coolMark
            ) --评定
            v.MARK = TowerMark
            --
            Ubertip = Ubertip .. "|cffccffcc阶级：" .. tlv .. "|r"
            Ubertip = Ubertip .. "|n|cffff0000攻击类型：" .. CONST_WEAPON_TYPE[v.weapTp1] .. "(" .. v.cool1 .. "秒/击)|r"
            Ubertip = Ubertip .. "|n|cffff8080攻击能力：" .. v.ATTACK_WHITE.."|r"
            if (Primary == "STR") then
                Ubertip = Ubertip .. "|n|cffffff00力量：" .. v.STR .. "(+" .. v.STRplus .. ") *|r"
            else
                Ubertip = Ubertip .. "|n|cffffffcc力量：" .. v.STR .. "(+" .. v.STRplus .. ")|r"
            end
            if (Primary == "AGI") then
                Ubertip = Ubertip .. "|n|cffffff00敏捷：" .. v.AGI .. "(+" .. v.AGIplus .. ") *|r"
            else
                Ubertip = Ubertip .. "|n|cffffffcc敏捷：" .. v.AGI .. "(+" .. v.AGIplus .. ")|r"
            end
            if (Primary == "INT") then
                Ubertip = Ubertip .. "|n|cffffff00智力：" .. v.INT .. "(+" .. v.INTplus .. ") *|r"
            else
                Ubertip = Ubertip .. "|n|cffffffcc智力：" .. v.INT .. "(+" .. v.INTplus .. ")|r"
            end
            Ubertip = Ubertip .. "|n|cffcc99ff评定："..TowerMark.."分|r"
            Ubertip = Ubertip .. "|n|n" .. hColor.grey(UberDesc)
            --塔基的属性说明
            local dobj = slk.ability.Aamk:new("towerOriginAbli_" .. thisIndex)
            local Name = "[兵塔参数]" .. v.Name
            local Tip = v.Name
            dobj.Name = Name
            dobj.Tip = Tip
            dobj.Ubertip = Ubertip
            dobj.Buttonpos1 = 0
            dobj.Buttonpos2 = 0
            dobj.hero = 0
            dobj.levels = 1
            dobj.DataA1 = 0
            dobj.DataB1 = 0
            dobj.DataC1 = 0
            dobj.Art = v.Art
            v.DESC_ABILITY_ID = dobj:get_id()
            --
            obj.Hotkey = ""
            obj.tilesets = 1
            obj.hostilePal = 0
            obj.Requires = "" --需求,全部无限制，用代码限制
            obj.Requirescount = 1
            obj.Requires1 = ""
            obj.Requires2 = ""
            obj.upgrade = ""
            obj.collision = 0.00
            obj.unitShadow = "ShadowFlyer"
            obj.Buttonpos1 = 0
            obj.Buttonpos2 = 0
            obj.death = 0.10
            obj.turnRate = 1.00
            obj.acquire = 749 + (v.RangeInc or 0)
            obj.weapsOn = 1
            obj.race = "human"
            obj.deathType = 0
            obj.fused = 0
            obj.sides1 = 5 --骰子面
            obj.dice1 = 1 --骰子数量
            obj.regenMana = 0.00
            obj.regenHP = 0.00
            obj.stockStart = 0
            obj.stockRegen = 0
            obj.stockMax = 1
            obj.collision = 16 --接触体积
            obj.def = v.def or 0.00 -- 护甲
            obj.sight = v.sight or 1000 -- 白天视野
            obj.nsight = v.nsight or 1000 -- 夜晚视野
            local targs1 = "vulnerable,ground,ward,structure,organic,mechanical,tree,debris,air" --攻击目标
            obj.targs1 = targs1
            obj.EditorSuffix = "#" .. tlv
            obj.Propernames = tlv .. "阶兵塔"
            local abl = {}
            if(type(v.abilList) == "string")then
                abl = string.explode(',', abl)
            elseif(type(v.abilList) == "table")then
                for _,t in pairs(v.abilList) do
                    table.insert( abl, t )
                end
            end
            if(#abl ==0)then
                abl = {
                    towerSpxKV["封印枷锁之一"],
                    towerSpxKV["封印枷锁之二"]
                }
            elseif(#abl == 1)then
                table.insert( abl, towerSpxKV["封印枷锁之二"] )
            end
            table.insert( abl, "AInv" )
            table.insert( abl, v.DESC_ABILITY_ID )
            table.insert( abl, unitPowerMap[tlv] )
            obj.abilList = string.implode(",",abl)
            obj.heroAbilList = ""
            obj.nameCount = v.nameCount or 1
            if (v.weapTp1 == "msplash" or v.weapTp1 == "artillery") then
                --溅射/炮火
                obj.Farea1 = v.Farea1 or 1
                obj.Qfact1 = v.Qfact1 or 0.05
                obj.Qarea1 = v.Qarea1 or 500
                obj.Hfact1 = v.Hfact1 or 0.15
                obj.Harea1 = v.Harea1 or 350
                obj.splashTargs1 = targs1 .. ",enemies"
            end
            if (v.weapTp1 == "mbounce") then
                --弹射
                obj.Farea1 = v.Farea1 or 450
                obj.targCount1 = v.targCount1 or 4
                obj.damageLoss1 = v.damageLoss1 or 0.3
                obj.splashTargs1 = targs1 .. ",enemies"
            end
            if (v.weapTp1 == "mline") then
                --穿透
                obj.spillRadius = v.spillRadius or 200
                obj.spillDist1 = v.spillDist1 or 450
                obj.damageLoss1 = v.damageLoss1 or 0.3
                obj.splashTargs1 = targs1 .. ",enemies"
            end
            if (v.weapTp1 == "aline") then
                --炮火穿透
                obj.Farea1 = v.Farea1 or 1
                obj.Qfact1 = v.Qfact1 or 0.05
                obj.Qarea1 = v.Qarea1 or 500
                obj.Hfact1 = v.Hfact1 or 0.15
                obj.Harea1 = v.Harea1 or 350
                obj.spillRadius = v.spillRadius or 200
                obj.spillDist1 = v.spillDist1 or 450
                obj.damageLoss1 = v.damageLoss1 or 0.3
                obj.splashTargs1 = targs1 .. ",enemies"
            end
            obj.Tip = "选择 " .. v.Name
            obj.Name = "["..tlv.."阶]" .. v.Name
            obj.Awakentip = "复活 " .. v.Name
            obj.Revivetip = "复活 " .. v.Name
            obj.Tip = "召唤 " .. v.Name
            obj.Ubertip = Ubertip
            obj.unitSound = v.unitSound -- 声音
            obj.modelScale = v.modelScale --模型缩放
            obj.file = v.file --模型
            obj.fileVerFlags = v.fileVerFlags or 0
            obj.Art = v.Art --头像
            obj.scale = v.scale or 1.00 --选择圈
            local movetp = v.movetp or "foot"
            local moveHeight = v.moveHeight or 0
            if(movetp == 'fly')then
                moveHeight = 250
            end
            obj.movetp = movetp --移动类型
            obj.moveHeight = moveHeight --移动高度
            obj.moveFloor = moveHeight * 0.25 --最低高度
            obj.spd = 100
            obj.backSw1 = v.backSw1 or 0.500
            obj.dmgpt1 = v.dmgpt1 or 0.500
            obj.rangeN1 = 750 + (v.RangeInc or 0)
            obj.cool1 = v.cool1 or 2.00
            obj.armor = "Flesh" -- 被击声音
            obj.targType = "ground" --作为目标类型
            obj.Missileart = v.Missileart -- 箭矢模型
            obj.Missilespeed = 1100 -- 箭矢速度
            obj.Missilearc = v.Missilearc or 0.05
            obj.weapTp1 = v.weapTp1 or "normal" --攻击类型
            obj.weapType1 = "" --攻击声音
            obj.Primary = Primary
            obj.dmgplus1 = 1 -- 基础攻击
            obj.showUI1 = 0 -- 不显示攻击按钮
            obj.STR = v.STR
            obj.AGI = v.AGI
            obj.INT = v.INT
            obj.STRplus = v.STRplus
            obj.AGIplus = v.AGIplus
            obj.INTplus = v.INTplus
            v.INDEX = thisIndex
            v.TOWER_POWER = tlv
            v.UNIT_ID = obj:get_id()
            -- 塔基物品
            local iobj = slk.item.gold:new("towers_items_" .. thisIndex)
            iobj.Name = "[" .. tlv .. "][" .. v.Name .. "]"
            iobj.Tip = "点击召唤兵塔：[" .. v.Name .. "]"
            iobj.UberTip = Ubertip
            iobj.Description = Ubertip
            iobj.Art = v.Art
            iobj.scale = 1.10
            iobj.selSize = 80
            iobj.goldcost = math.floor(TowerMark * 0.7)
            iobj.lumbercost = 0
            iobj.sellable = 1
            iobj.cooldownID = UsedID[2][2]
            iobj.file = "war3mapImported\\item_Stone.mdl"
            iobj.abilList = UsedID[2][2]
            iobj.perishable = 1
            iobj.powerup = 0
            local hitem = {
                INDEX = thisIndex,
                Name = v.Name,
                Art = v.Art,
                goldcost = 0,
                lumbercost = 0,
                perishable = 1,
                powerup = 0,
                WEIGHT = 0,
                OVERLIE = 999,
                ITEM_ID = iobj:get_id(),
                UNIT_ID = v.unitID,
                TOWER_POWER = tlv,
            }
            ?>
            call SaveStr(hash_myslk, StringHash("towers"), <?=towersTi?>, "<?=string.addslashes(json.stringify(v))?>")
            call SaveStr(hash_myslk, StringHash("towersItems"), <?=towersTi?>, "<?=string.addslashes(json.stringify(hitem))?>")
            <?
            --shadow
            createTowerShadowUnit(v,towersTi,tlv)
        end
    end
end

?>
call SaveInteger(hash_myslk, StringHash("towers"), -1, <?=towersTi?>)
<?